File: src/main/java/org/noxet/noxetserver/messaging/Message.java
package org.noxet.noxetserver.messaging;

import net.md_5.bungee.api.ChatColor;
import net.md_5.bungee.api.ChatMessageType;
import net.md_5.bungee.api.chat.ClickEvent;
import net.md_5.bungee.api.chat.HoverEvent;
import net.md_5.bungee.api.chat.TextComponent;
import net.md_5.bungee.api.chat.hover.content.Text;
import org.bukkit.World;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Player;
import org.noxet.noxetserver.NoxetServer;
import org.noxet.noxetserver.RealmManager;
import org.noxet.noxetserver.messaging.channels.MessagingChannel;
import org.noxet.noxetserver.util.TextBeautifier;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

public class Message {
    private final List<TextComponent> textComponents = new ArrayList<>();
    private String prefix = getDefaultPrefix();
    protected ChatMessageType chatMessageType = ChatMessageType.CHAT;

    /**
     * Constructs a message with a text.
     * @param text The text message to be sent
     */
    public Message(String text) {
        add(text, null, null);
    }

    public String getPrefix() {
        return prefix;
    }

    public void setPrefix(String newPrefix) {
        prefix = newPrefix;
    }

    public static String getDefaultPrefix() {
        return "§b" + TextBeautifier.beautify("no") + "§3§lx§b" + TextBeautifier.beautify("et") + " §8§l| ";
    }

    protected TextComponent getBakedMessage() {
        TextComponent mainComponent = new TextComponent(prefix != null ? prefix : "");

        for(TextComponent textComponent : textComponents)
            mainComponent.addExtra(textComponent); // Add every component.

        return mainComponent;
    }

    public ChatColor getDefaultColor() {
        return ChatColor.GRAY;
    }

    public Message add(String text, String hoverText, String clickCommand) {
        if(text == null)
            return this;

        TextComponent textComponent = new TextComponent(text + "§r "); // Padding, for better transitions.

        textComponent.setColor(getDefaultColor());

        if(hoverText != null)
            textComponent.setHoverEvent(new HoverEvent(HoverEvent.Action.SHOW_TEXT, new Text(hoverText)));

        if(clickCommand != null)
            textComponent.setClickEvent(new ClickEvent(ClickEvent.Action.RUN_COMMAND, "/" + clickCommand));

        textComponents.add(textComponent);

        return this;
    }

    public Message add(String text, String hoverText) {
        return add(text, hoverText, null);
    }

    public Message add(String text) {
        return add(text, null, null);
    }

    public Message addButton(String buttonLabel, ChatColor color, String hoverText, String clickCommand) {
        TextComponent textComponent = new TextComponent("→" + TextBeautifier.beautify(buttonLabel) + " ");

        textComponent.setColor(color);
        textComponent.setBold(true);

        if(hoverText != null)
            textComponent.setHoverEvent(new HoverEvent(HoverEvent.Action.SHOW_TEXT, new Text(hoverText)));

        if(clickCommand != null)
            textComponent.setClickEvent(new ClickEvent(ClickEvent.Action.RUN_COMMAND, "/" + clickCommand));

        textComponents.add(textComponent);

        return this;
    }

    public void send(Player player) {
        player.spigot().sendMessage(chatMessageType, getBakedMessage());
    }

    public void send(CommandSender commandSender) {
        commandSender.spigot().sendMessage(getBakedMessage());
    }

    public void send(Collection<? extends Player> players) {
        for(Player player : players)
            send(player);
    }

    public void send(World world) {
        send(world.getPlayers());
    }

    public void send(World[] worlds) {
        for(World world : worlds)
            send(world);
    }

    public void send(RealmManager.Realm realm) {
        if(realm != null)
            send(realm.getWorlds().toArray(new World[0]));
        else
            send(NoxetServer.ServerWorld.HUB.getWorld());
    }

    public void send(MessagingChannel channel) {
        for(CommandSender recipient : channel.getRecipients())
            send(recipient);
    }

    public void broadcast() {
        Collection<? extends Player> sendTo = NoxetServer.getPlugin().getServer().getOnlinePlayers();
        send(sendTo);
    }
}

File: src/main/java/org/noxet/noxetserver/messaging/Message.scala
package org.noxet.noxetserver.messaging

import net.md_5.bungee.api.ChatColor
import net.md_5.bungee.api.ChatMessageType
import net.md_5.bungee.api.chat.ClickEvent
import net.md_5.bungee.api.chat.HoverEvent
import net.md_5.bungee.api.chat.TextComponent
import net.md_5.bungee.api.chat.hover.content.Text
import org.bukkit.World
import org.bukkit.command.CommandSender
import org.bukkit.entity.Player
import org.noxet.noxetserver.NoxetServer
import org.noxet.noxetserver.realm.RealmManager.Realm
import org.noxet.noxetserver.messaging.channels.MessagingChannel
import org.noxet.noxetserver.util.TextBeautifier

import scala.collection.mutable.ListBuffer
import scala.jdk.CollectionConverters.*
import scala.collection.Seq

object Message:
  private val defaultPrefix = s"§b${TextBeautifier.beautify("no")}§3§lx§b${TextBeautifier.beautify("et")} §8§l| "

class Message:
  private val textComponents = ListBuffer[TextComponent]()
  private var prefix: Option[String] = Some(Message.defaultPrefix)
  protected var chatMessageType: ChatMessageType = ChatMessageType.CHAT

  /**
   * Constructs a message with a text.
   *
   * @param text The text message to be sent
   */
  def this(text: String) =
    this()
    add(text)

  def getDefaultColor: ChatColor = ChatColor.GRAY

  // Make prefix public and remove getters and setters?
  def getPrefix: String = prefix.get

  def setPrefix(newPrefix: Option[String]): Unit = prefix = newPrefix

  protected def bake: TextComponent =
    val mainComponent = TextComponent(prefix.getOrElse(""))

    textComponents.foreach(mainComponent.addExtra)

    mainComponent

  def add(text: String, hoverText: Option[String] = None, clickCommand: Option[String] = None): Message =
    if text == null then return this // TODO ????

    val textComponent = TextComponent(s"$text§r "); // Padding, for better transitions.

    textComponent.setColor(getDefaultColor)

    hoverText.foreach(v => textComponent.setHoverEvent(HoverEvent(HoverEvent.Action.SHOW_TEXT, Text(v))))
    clickCommand.foreach(v => textComponent.setClickEvent(ClickEvent(ClickEvent.Action.RUN_COMMAND, s"/$v")))

    textComponents.addOne(textComponent)

    this

  def addButton(label: String, color: ChatColor, hoverText: Option[String], clickCommand: Option[String]): Message =
    val textComponent = TextComponent(s"→${TextBeautifier.beautify(label)} ")

    textComponent.setColor(color)
    textComponent.setBold(true)

    hoverText.foreach(v => textComponent.setHoverEvent(HoverEvent(HoverEvent.Action.SHOW_TEXT, Text(v))))
    clickCommand.foreach(v => textComponent.setClickEvent(ClickEvent(ClickEvent.Action.RUN_COMMAND, s"/$v")))

    textComponents.addOne(textComponent)

    this

  def send(player: Player): Unit = player.spigot.sendMessage(chatMessageType, bake)

  def send(commandSender: CommandSender): Unit = commandSender match
    case p: Player => send(p)
    case _ => commandSender.spigot.sendMessage(bake)

  def send(players: Seq[Player]): Unit = players.foreach(send)

  def send(world: World): Unit = send(world.getPlayers.asScala)

  def send(worlds: => Seq[World]): Unit = worlds.foreach(send)

  def send(realm: Realm): Unit =
    if realm != null then send(realm.getWorlds.asScala)
    else send(NoxetServer.ServerWorld.HUB.getWorld)

  def send(channel: MessagingChannel): Unit = channel.getRecipients.forEach(send)

  def broadcast(): Unit = send(NoxetServer.getPlugin.getServer.getOnlinePlayers.asScala.toSeq)